### ISSUE TITLE
[feature] Allow GTS server operators to configure pre-authorized OAuth applications by exact match  on redirect URI 

### ISSUE THREAD

**Issue by ForestJohnson** [superseriousbusiness/gotosocial#313]
GTS is great but it needs a user interface in order to be approachable by every-day users. I believe that most folks aren't going to understand the concept of the server being separate from the client.  It may be common in tech savvy circles (matrix users and ActivityPub power users)   but for the "soccer moms"  I think its currently too much.  

So I propose a new configuration option  `GTS_SKIP_OAUTH_CONSENT_FOR_REDIRECT_URLS`.  open  to other ideas about what to name it. It would be a comma separated list of URLs, default value would be emptystring

So for example if I host pinafore on `social.sequentialread.com` and I host GTS on `gotosocial.sequentialread.com` then I can configure 

```
GTS_SKIP_OAUTH_CONSENT_FOR_REDIRECT_URLS=https://social.sequentialread.com/settings/instances/add
```

This way I can tell new folks who want to sign up for an account to simply visit https://social.sequentialread.com/

And the following flow will ensue:

  
  1.The person will click on the link to https://social.sequentialread.com/ and their browser will load the pinafore home page
  
  2. I have modified pinafore with a small hack so that if the browser isn't logged in yet, it will automatically launch the login flow  for a configurable server, in this case the server gotosocial.sequentialread.com
  
  3. the user gets instantly redirected to gotosocial.sequentialread.com/auth/sign_in
  
  4. if the user already has an account, they log in. otherwise, they click on a  "Don't have an account yet? click here to register!" link.
  
        * note that we are still inside the oauth flow, so this registration page link has to carry that oauth session with it.    
        
  5. after the user has either logged in or registered w/ GTS, they will be redirected to the "add new oauth app" handler 
  
  6. this handler will check the redirect URL of the current "new oauth app" session  against the list in the configuration. if it does not find an exact match, it will display the consent screen to the user as normal ("Are you sure you want to allow pinafore.social to access your account?").  Otherwise if it DOES find an exact match, it will skip the consent screen and automatically approve it, directly re-directing to the redirect URL w/ the token.

The user is already going to think of pinafore and GTS as the same app anyways, so I feel like we might as well allow GTS server operators to help users maintain this mental model and provide a user experience that is relatable to centralized platforms that folks have used in the past.   They can still dig into it and learn about how to connect mobile apps, etc, and we can provide help for that thru small modifications to the Pinafore UI, but I think it's important to start out with an onboarding process that is only built out of interactions pieces that everyone has seen before.

**Comment by ForestJohnson** [comment]
I'm fairly sure this would be a secure configuration but I'm still slightly new to this stuff.

My thought process was in order to exploit  the consent screen skip, you would have to own or control the page at the redirect URI.  And in order to do that you would have to either be the server owner, in which case its fine as you own all the users data anyways, or you would have to execute XSS against that page. But if you get XSS on pinafore then you already own the user's account anyways.

Do we need to specify a set of scopes that are allowed for this consent screen skip?  Should it just default to only allowing the scopes that are requested by a typical client like pinafore, or does it need to be configurable? and does it need to be configurable per-URL?

**Comment by tsmethurst** [comment]
Being able to skip to permissions screen for 'trusted' redirect URIs is a good idea. Probably something like `GTS_TRUSTED_REDIRECT_URIS` would be a good name for it, or `GTS_TRUSTED_OAUTH_REDIRECTS`. This will become especially relevant if people start writing custom client applications that use GtS as a backend for whatever they want their app to do.

> note that we are still inside the oauth flow, so this registration page link has to carry that oauth session with it

That makes sense, and I think it's something we can do with the session implementation we currently have. We're going to have to do some juggling to differentiate between new registrations that are *just* using the GtS registration page, and new registrations that originated from an attempted client login.

**Comment by tsmethurst** [comment]
> Do we need to specify a set of scopes that are allowed for this consent screen skip?

I don't think so... if we're 'trusting' the app, let's just allow it the scopes that it requested, imo.

**Comment by daenney** [comment]
We've migrated to Codeberg. Your issue has been imported as https://codeberg.org/superseriousbusiness/gotosocial/issues/313.
